File: snapshot/snapshot.py
#!/usr/bin/python

from epics import caget
import time
import json
import sys

if len(sys.argv) == 1:
    config = open('./config/pv_list.json', 'r')
elif len(sys.argv) == 2:
    if str(sys.argv[1]) == 'clean':
        config = open('./config/pv_list_clean.json', 'r')
    else:
        config = open('./config/pv_list.json', 'r')
else:
    sys.exit('at most two arguments allowed.')
result = {}
pv_list = json.loads(config.read())
start = time.time()
for pv in pv_list:
    result[pv] = caget(pv, timeout=1)
duration = time.time() - start
print int(round(duration * 1000))

File: snapshot/snapshot.js
var ca = require(__dirname + '/lib/ca.js'),
  fs = require('fs'),
  pv_list,
  size,
  results = {},
  complete = 0,
  start = Date.now();

if (process.argv.length == 2) {
  pv_list = JSON.parse(fs.readFileSync(__dirname + '/config/pv_list.json'));
} else if (process.argv.length == 3) {
  if (process.argv[2] === 'clean') {
    pv_list = JSON.parse(fs.readFileSync(__dirname + '/config/pv_list_clean.json'));
  } else {
    pv_list = JSON.parse(fs.readFileSync(__dirname + '/config/pv_list.json'));
  }
} else {
  console.warn('too many arguments.');
  return;
}

size = pv_list.length;

pv_list.forEach(function(pv) {
  ca.exec('caget', pv, function(err, result) {
    complete = complete + 1;
    if (err) {
      results[pv] = {
        name: pv,
        value: 'unavailable'
      };
    } else {
      results[pv] = ca.parseCaget(result);
    }
    if (complete == size) {
      console.log('the snapshot took ' + (Date.now() - start) + ' milliseconds.');
    }
  });
});



